go-controller/pkg/metrics/ovnkube_controller.go

```diff
@@ -619,6 +647,106 @@ func ipsecMetricHandler(table string, model model.Model) {
 	}
 }
 
+// MonitorIPsecTunnelsState will register a labeled metric for the ipsec tunnel status. It will subscribe
+// to xfrm netlink events and update the metric with the tunnel up or down status.
+func MonitorIPsecTunnelsState(stopChan <-chan struct{}, wg *sync.WaitGroup, ovsVsctl ovsClient, ipsec ipsecClient) {
+	prometheus.MustRegister(metricIPsecTunnelState)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		groups, err := xfrmMcastGroups([]nl.XfrmMsgType{nl.XFRM_MSG_NEWSA, nl.XFRM_MSG_DELSA, nl.XFRM_MSG_UPDSA,
+			nl.XFRM_MSG_NEWPOLICY, nl.XFRM_MSG_DELPOLICY, nl.XFRM_MSG_UPDPOLICY})
+		if err != nil {
+			klog.Errorf("Error creating xfrm Mcast groups, err: %v", err)
+			return
+		}
+		sock, err := nl.SubscribeAt(netns.None(), netns.None(), unix.NETLINK_XFRM, groups...)
+		if err != nil {
+			klog.Errorf("Error subscribing xfrm events, err: %v", err)
+			return
+		}
+		if stopChan != nil {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				<-stopChan
+				sock.Close()
+			}()
+		}
+		for {
+			msgs, from, err := sock.Receive()
```

Comment:
how does this end if stop signal received? Its a bit strange and hard to read that youve another go routine above and then this for loop below that seems to do its own thing.. i see itll probably emit an info log but this is weird.. i think you should use the common pattern we do - switch statement